### DISCUSSION TITLE
how to get final load listing

### DISCUSSION THREAD

**Discussion by pm100** [cc65/cc65#2273] [Q&A]
using assemble and link I cannot find a way to get a listing showing the source and the actual load address. I have to get the verbose map from the linker and then look at the assembler listing, work out which segment something is in then add the base of that segment from each address (or vice versa to get the source address from a runtime address)

Is there a way to get a listing showing the final load address of each instruction?

ie this listing
```

001DECr 2               .segment "INIT"
000000r 2               
000000r 2               .ifdef KBD
000000r 2               FNDLIN2:
000000r 2                       php
000000r 2                       jmp     FNDLIN
000000r 2               .endif
000000r 2               
000000r 2               ; ----------------------------------------------------------------------------
000000r 2               PR_WRITTEN_BY:
000000r 2               .ifndef KBD
000000r 2                 .ifndef CONFIG_CBM_ALL
000000r 2  A9 rr                lda     #<QT_WRITTEN_BY
000002r 2  A0 rr                ldy     #>QT_WRITTEN_BY
000004r 2  20 rr rr             jsr     STROUT
```

shows code in the INIT segment

map says

`INIT                  004062  004264  000203  00001`

so I know that the first LDA is actually at 0x4062. It would be great to have 

```
004062  A9 rr                lda     #<QT_WRITTEN_BY
004064  A0 rr                ldy     #>QT_WRITTEN_BY
```

Having the 'rr's replaced too would be nice but less important

**Comment by silverdr** [comment]
Actual load addresses are not known until linking, and that's a feature ;-)

**Comment by pm100** [comment]
I know, but after assemble and link they are known. There could be a tool which produces a final listing by reading the map file generated by the linker and the list file from the assembler

You can imagine a version of ca65 that reads the map file to produce an annotated absolute listing, it would be a second pass.

having the assembler list addresses as seg.offset would be useful too

**Reply by silverdr** [reply]
ca65 is an assembler - it is not supposed to know anything about linking or what artefacts linker is supposed to leave behind. If anything then possibly the "compile and link" utility could take such extra task - IMHO

**Comment by spiro-trikaliotis** [comment]
This is not completely out of scope.

In 2021, I started this. My idea was to generate a kind of "post-processing" of the assembler listing files of ca65 after the linker step, so the addresses could be adjusted.

However, I did not go very far:

1. The ca65 listing is missing expanded macros (and other curcial parts); that is, you do not get the whole contents of the file being assembled, which makes it hard to follow once there is a macro over a specific limit. See also #1427
3. Thus, I started an option to create a full listing file. It is not ready for merge yet, but I once had an incomplete preview here: #1436.
4. While working on it, I started adding test cases for the ca65 assembler, because I found that this part if the code is prone to adding regressions. (cf. #1679, #1771,  #1916)

So, there was much distraction and not so much progress because of lack in time. The patches above are not even all that is there; I  added some work. Unfortunately, someone started working on ca65 and made changes that completely broke my patches in a way that it takes much time to regenerate them again.

That's when I lost interest in these patches. However, I am still convinced it is doable.

**Reply by pm100** [reply]
You and I clearly have a similar experience with this. The msbasic 6502 port is full of macros that generate code into 2 segments, it's really hard to follow the output. It has a lot of segments too, scrolling back to find which one you are in is time consuming. I think 1.00001 or 2.02345 would be a good output syntax. Ie seg number and offset

**Reply by silverdr** [reply]
How about simply post-processing the current artefacts without changing anything in the assembler and linker? Or do those lack some information pieces to make it work? Or maybe they contain some bits, which make this infeasible? Otherwise I could imagine this kicked-off as a separate utility, which upon maturing could be incorporated into `cl65`?

**Reply by pm100** [reply]
That was one of my suggestions. It would involve reading the human readable text output of ca65. I never really like doing that as it is fragile. Hence my idea of making a tool that's actually ca65 in disguise

**Reply by silverdr** [reply]
Yeah, nobody likes parsing formats, which may change at any time but unless there are problems with the information stored in those files, that seems like the easiest way forward. And if it's a part of the project then it can possibly be really stable as changes to the output from assembler OR linker would have to take this into account. The only approach I'd personally oppose would be to try to couple it directly with the assembler

**Comment by spiro-trikaliotis** [comment]
> How about simply post-processing the current artefacts without changing anything in the assembler and linker? Or do those lack some information pieces to make it work?

Yes, they lack data. There is info missing in the listing file (for example, the macros; first, all bytes of the macro are crushed into one long line. If a segment changes, the assembler listing does not account for that. Thus, it is not possible to follow the bytes in the listing file. That's why I wanted to create the "complete" listing for a start.


> Otherwise I could imagine this kicked-off as a separate utility, which upon maturing could be incorporated into `cl65`?

This is what I had in mind, too. I don't know if it makes send to integrate it into cl65 or not, but this can be argued.

I even got a perl (?) script from someone that works in some cases, but not if you use macros are other features where the ca65 listing omits important info. I promised not to share this script, though. (And I would have to dig it up, it is somewhere on my HDD).

**Reply by pm100** [reply]
Does this project accept PRs. First thing would be to add a flag to expand macros. That a separate PR. Second would be to include seg I'd in the address. That's also a free standing PR. Happy to do both but only in this project accepts PRs

**Reply by spiro-trikaliotis** [reply]
> Does this project accept PRs.

It does.

> First thing would be to add a flag to expand macros. That a separate PR.

That's not trivial.

When a macro is defined, ca65 only stores the tokens. There is not reference to the source line anymore. Thus, it cannot be output when the macro is expanded.

Because of this, in the patch I cited above, I started to store the lines alongside with the tokens, so they can be re-created. This is not trivial, though, because there are so many edge-cases. Because of these edge-cases, I started with the test cases, which did not exist until then.

> Second would be to include seg I'd in the address. That's also a free standing PR.

This could be possible, but I am not sure if this is really needed once the complete listing is there. It would make it easier to post-process, but not needed, IMHO.
